File: sdks/typescript/tests/interop.test.ts
import { readFileSync } from 'node:fs';
import { describe, expect, it } from 'vitest';
import { leafHash, verifyInclusion, verifyTreeHead } from '../src/verifier.js';
import type { SignedTreeHead } from '../src/types.js';

const INTEROP_DATA = process.env.INTEROP_DATA ?? '';

interface InteropEntry {
  seq: number;
  data_hex: string;
  leaf_hash_hex: string;
}

interface InteropSth {
  tree_size: number;
  root_hash_hex: string;
  timestamp_ns: number;
  timestamp_ns_str: string;
  signature_hex: string;
  public_key_hex: string;
  key_version: number;
}

interface InteropProof {
  seq: number;
  leaf_hash_hex: string;
  proof_hex: string[];
}

interface InteropData {
  entries: InteropEntry[];
  sth: InteropSth;
  inclusion_proofs: InteropProof[];
}

function hex(s: string): Uint8Array {
  return s.length === 0 ? new Uint8Array(0) : new Uint8Array(Buffer.from(s, 'hex'));
}

// Parsed once at module load; null when INTEROP_DATA is not set.
const d: InteropData | null = INTEROP_DATA
  ? (JSON.parse(readFileSync(INTEROP_DATA, 'utf-8')) as InteropData)
  : null;

describe.skipIf(!d)('cross-SDK interop', () => {
  it('STH signature verifies', () => {
    const sth: SignedTreeHead = {
      treeSize: BigInt(d!.sth.tree_size),
      rootHash: hex(d!.sth.root_hash_hex),
      // Use the string field to avoid float64 precision loss on large timestamps.
      timestampNs: BigInt(d!.sth.timestamp_ns_str),
      signature: hex(d!.sth.signature_hex),
      publicKey: hex(d!.sth.public_key_hex),
      keyVersion: d!.sth.key_version,
    };
    expect(verifyTreeHead(sth)).toBe(true);
  });

  for (const ip of d?.inclusion_proofs ?? []) {
    it(`inclusion proof seq ${ip.seq}`, () => {
      expect(
        verifyInclusion(
          hex(ip.leaf_hash_hex),
          BigInt(ip.seq),
          BigInt(d!.sth.tree_size),
          ip.proof_hex.map(hex),
          hex(d!.sth.root_hash_hex),
        ),
      ).toBe(true);
    });
  }

  for (const entry of d?.entries ?? []) {
    it(`leaf hash seq ${entry.seq}`, () => {
      const got = leafHash(hex(entry.data_hex));
      expect(Buffer.from(got).toString('hex')).toBe(entry.leaf_hash_hex);
    });
  }
});

File: sdks/rust/tests/interop_test.rs
use settled_sdk::verifier::{leaf_hash, verify_inclusion, verify_tree_head};

#[derive(serde::Deserialize)]
struct InteropEntry {
    seq: u64,
    data_hex: String,
    leaf_hash_hex: String,
}

#[derive(serde::Deserialize)]
struct InteropSth {
    tree_size: u64,
    root_hash_hex: String,
    timestamp_ns: i64,
    signature_hex: String,
    public_key_hex: String,
}

#[derive(serde::Deserialize)]
struct InteropProof {
    seq: u64,
    leaf_hash_hex: String,
    proof_hex: Vec<String>,
}

#[derive(serde::Deserialize)]
struct InteropData {
    entries: Vec<InteropEntry>,
    sth: InteropSth,
    inclusion_proofs: Vec<InteropProof>,
}

fn b32(s: &str) -> [u8; 32] {
    hex::decode(s).unwrap().try_into().expect("32 bytes")
}

fn hx(s: &str) -> Vec<u8> {
    hex::decode(s).unwrap()
}

#[test]
fn interop_verify() {
    let path = match std::env::var("INTEROP_DATA") {
        Ok(p) => p,
        Err(_) => return, // skip when not set
    };

    let raw = std::fs::read(&path)
        .unwrap_or_else(|e| panic!("cannot read {path}: {e}"));
    let d: InteropData = serde_json::from_slice(&raw).expect("parse interop data");

    assert!(
        verify_tree_head(
            d.sth.tree_size,
            b32(&d.sth.root_hash_hex),
            d.sth.timestamp_ns,
            &hx(&d.sth.signature_hex),
            &hx(&d.sth.public_key_hex),
        ),
        "STH signature verification failed"
    );

    for ip in &d.inclusion_proofs {
        let proof: Vec<[u8; 32]> = ip.proof_hex.iter().map(|s| b32(s)).collect();
        assert!(
            verify_inclusion(
                b32(&ip.leaf_hash_hex),
                ip.seq,
                d.sth.tree_size,
                &proof,
                b32(&d.sth.root_hash_hex),
            ),
            "inclusion proof for seq {} failed",
            ip.seq
        );
    }

    for entry in &d.entries {
        let data = hex::decode(&entry.data_hex).unwrap();
        let got = leaf_hash(&data);
        let want = b32(&entry.leaf_hash_hex);
        assert_eq!(got, want, "leaf hash mismatch for seq {}", entry.seq);
    }
}
